### PR TITLE
Manage members modal

### PR DESCRIPTION
**Pull request by ahsu1230**
@andrewconner @2is10

### CODE REVIEW

**Comment by ahsu1230** [comment]
Button "Manage/View members" appears on upper right corner of library header only if the library has any collaborators or followers (only if you're on the `collaborative` experiment)

File: kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+		<defs>
+			<path id="SVGID_3_" d="M13.1,7c-0.3,0-0.5,0-0.9,0.1c1,0.7,1.7,1.7,1.7,2.9v2.1h5.2V10C19.1,8,15.1,7,13.1,7z M6.2,7
+				c-2,0-6,1-6,3v2.1h12.1V10C12.2,8,8.2,7,6.2,7z M6.2,5.3c1.5,0,2.6-1.1,2.6-2.6c0-1.5-1.1-2.6-2.6-2.6S3.6,1.2,3.6,2.7
+				C3.6,4.2,4.7,5.3,6.2,5.3z M13.1,5.3c1.5,0,2.6-1.1,2.6-2.6c0-1.5-1.1-2.6-2.6-2.6c-1.5,0-2.6,1.1-2.6,2.6
+				C10.5,4.2,11.6,5.3,13.1,5.3z"/>
+		</defs>
+		<clipPath id="SVGID_4_">
+			<use xlink:href="#SVGID_3_"  overflow="visible"/>
+		</clipPath>
+		<rect x="-4.9" y="-4.9" clip-path="url(#SVGID_4_)" fill="#73C385" width="29" height="22"/>
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg]: I’ve updated the `collaborate-white.svg` and `collaborate-gray.svg` files to simplify them (just a single `<path>` element), without altering how they are rendered. Would you mind just copying one of those instead and changing the fill color to the color in this file (`#73c385`)?

Also, let’s be consistent in the file names (“collaborator” or “collaborate”). Thanks.

File: kifi-backend/modules/shoebox/angular/src/common/services/net.js [hunk 2]
```diff
@@ -23,6 +23,8 @@ angular.module('kifi')
       removeKeepFromLibrary: del(shoebox, '/libraries/:id/keeps/:keepId'),
       removeManyKeepsFromLibrary: post(shoebox, '/libraries/:id/keeps/delete'),
 
+      updateLibraryMembership: post(shoebox, '/libraries/:id/members/:uid/access'),
+
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/common/services/net.js:27]: Nice!

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryHeader.tpl.html [hunk 3]
```diff
@@ -41,6 +41,13 @@
           <span class="kf-lh-followers-more-text kf-{{numMoreFollowersText.length}}">{{numMoreFollowersText}}</span>
         </button>
       </div>
+
+      <div class="kf-lh-members" ng-if="onCollabExperiment && (library.numCollaborators > 0 || library.numFollowers > 0)">
+        <button class="kf-lh-view-members" ng-click="openMembersModal()">
+          {{(amOwner || (amCollab && collabsCanEdit)) ? 'Manage members' : 'View members'}}
+        </button>
+      </div>
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryHeader.tpl.html:49]: This button is going to replace the follower pictures for all libraries—even non-collaborative ones? That’ll make a lot of library pages look a lot less interesting. Just making sure that’s the intent.

Comment (ahsu1230) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryHeader.tpl.html:49]: Yea, according to the design

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                member.isCollaborator = isCollaborator(member);
+                member.isFollower = isFollower(member);
+              });
+              loading = false;
+              if (members.length === 0) {
+                scope.moreMembers = false;
+              } else {
+                scope.moreMembers = true;
+                scope.offset += 1;
+                members = _.reject(members, function(m) { return m.lastInvitedAt; });
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js]: Or:

``` js
_.remove(members, 'lastInvitedAt');
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                scope.memberList.push.apply(scope.memberList, members);
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js:52]: Nice!

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+        function updateMembership(member, access) {
+          scope.selectedMemberId = member.id;
+          return net.updateLibraryMembership(scope.library.id, member.id, {access: access});
+        }
+
+        function countFollowersAndCollaborators() {
+          var memCounts = _.countBy(scope.memberList, function(mem) {
+            return mem.membership;
+          });
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js]: Nice! Shorter:

``` js
var memCounts = _.countBy(scope.memberList, 'membership');
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+          scope.library.numFollowers = memCounts[followerAccess] || 0;
+          scope.library.numCollaborators = memCounts[collabAccess] || 0;
+        }
+
+        scope.toggleWhoCanInvite = function() {
+          if (scope.collabCanInvite) {
+            libraryService.modifyLibrary({'id': scope.library.id, 'whoCanInvite': ownerOnlyInviteSetting}, false);
+          } else {
+            libraryService.modifyLibrary({'id': scope.library.id, 'whoCanInvite': collabInviteSetting}, false);
+          }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js]: Would be better without the code duplication:

``` js
libraryService.modifyLibrary({
  'id': scope.library.id,
  'whoCanInvite': scope.collabCanInvite ? ownerOnlyInviteSetting : collabInviteSetting
}, false);
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        };
+
+        scope.changeToCollaborator = function(member) {
+          updateMembership(member, collabAccess).then(function() {
+            var index = _.indexOf(scope.memberList, _.find(scope.memberList, {id: scope.selectedMemberId}));
+            scope.memberList[index].membership = collabAccess;
+            scope.memberList[index].isCollaborator = isCollaborator(scope.memberList[index]);
+            scope.memberList[index].isFollower = isFollower(scope.memberList[index]);
+            countFollowersAndCollaborators();
+          });
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js:125]: This callback function is identical to the one above inside `changeToFollower`. Consider factoring out and reusing.

Also, `index` seems irrelevant here. Consider:

``` js
var m = _.find(scope.memberList, {id: scope.selectedMemberId});
m.membership = collabAccess;
m.isCollaborator = isCollaborator(m);
m.isFollower = isFollower(m);
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        };
+
+        scope.removeMember = function(member) {
+          updateMembership(member, noAccess).then(function () {
+            $timeout(function() {
+              scope.memberList = _.drop(scope.memberList, {id: scope.selectedMemberId});
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.js]: Or:

``` js
_.remove(scope.memberList, {id: scope.selectedMemberId});
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.styl [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  cursor: pointer
+
+  .svg-dropdown-arrow
+    width: 16px
+    height: 8px
+    background-size: cover
+
+  .kf-dropdown-menu
+    dur = .1s
+    top: 'calc(100% + %s)' % actionGapBelow
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.styl:90]: Hmm, I don’t think this Stylus constant (`actionGapBelow`) is defined in this file.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.tpl.html [hunk 11]
```diff
@@ -0,0 +1,65 @@
+<div class="library-members dialog-content">
+  <a class="dialog-x" ng-click="close()"></a>
+
+  <div class="dialog-body-wrap">
+    <h1 class="dialog-header" ng-if="library.numFollowers > 0 || library.numCollaborators > 0">Manage members for this library</h1>
+    <h1 class="dialog-header" ng-if="library.numFollowers === 0 && library.numCollaborators === 0">{{library.name}}</h1>
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembers.tpl.html]: Rather than duplicate the element, consider using `?` `:` inside it:

``` HTML
<h1 class="dialog-header">
 {{library.numFollowers > 0 || library.numCollaborators > 0 ? 'Manage members for this library' : library.name}}
</h1>
```

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryMembersModal.tpl.html [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<div kf-modal class="kf-library-members" kf-width="420px">
+  <div kf-library-members></div>
+</div>
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryMembersModal.tpl.html:3]: I’ve noticed this convention before, of having small modal templates that use a different directive for their content. Do you know how the convention began? I can see it being useful if the content needs to be reusable outside of a modal (like persona selection), but I think that kind of reuse is generally rare. No need to change it here. Just curious.

File: kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg (hunk 1 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg]: ah so it seems like `collaborate-white` and `collaborate-gray` are different icons than `collaborator-green`. The first 2 have a plus sign next to the icon to indicate adding collaborators.
This green icon doesn't have the plus sign.

Should we rename the white & gray to something like `add-collaborate-white.svg`?

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 13]
```diff
@@ -197,25 +197,28 @@ angular.module('kifi')
...
-          return $q.reject({'error': 'missing fields: ' + missingFields.join(', ')});
+      modifyLibrary: function (opts, checkMissingFields) {
+        if (checkMissingFields) {
+          var required = ['name', 'visibility', 'slug'];
+          var missingFields = _.filter(required, function (v) {
+            return opts[v] === undefined;
+          });
+          if (missingFields.length > 0) {
+            return $q.reject({'error': 'missing fields: ' + missingFields.join(', ')});
+          }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:221]: I see that you’ve updated the calls in manageLibrary.js but it looks like there’s another call to `.createLibrary` in keepToLibraryWidget.js.

File: kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg (hunk 1 above)
Comment (2is10) [inline kifi-backend/modules/shoebox/angular/img/svg/collaborator-green.svg]: Oh, I see. It’s fine for this one to have a different name, then. Maybe `collaborators-green.svg`? (since it shows two heads)

**Comment by 2is10** [comment]
Nice work. Merge when you’re happy with it.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js (hunk 13 above)
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:221]: ah woops messed that!

**Comment by ahsu1230** [comment]
Thanks for the review & comments! @2is10 I always forget about those lodash shorthands! >.<
